### DISCUSSION TITLE
Struggling with overriding and templates

### DISCUSSION THREAD

**Discussion by MattF-NSIDC** [cue-lang/cue#1224] [Q&A] [answered]
Hello all, my team is working on a project called QGreenland (https://github.com/nsidc/qgreenland), and we're looking for ways to simplify our ~6000 lines of static YAML (enhanced with Python code to de-reference relationships and interpolate templates). We're just in the evaluation phase and comparing tools in this space. We admittedly are having some trouble grokking some concepts of the CUE language, but like that it is typed, as we're writing typed Python for the main codebase.

What we're struggling with is the use of templates and overrides at the same time.

This works as we expected:

```
$ cat test_template.cue
things: [Id=_]: {
  id: Id
  bar: string
}

things: {
  thing: {
    bar: "bar"
  }
}

$ cue eval -c test_template.cue 
things: {
    thing: {
        id:  "thing"
        bar: "bar"
    }
}
```

But we want each thing to have children, and each child can be one of many types. Here's what we feel is our best attempt so far. We can kind of see how it's wrong from the error message, but what we don't see is how to fix it :) Thank you for any help.

```
$ cat test.cue
#ChildThingA: #ChildThing & {
  attr_a: string
}
#ChildThingB: #ChildThing & {
  attr_b: string
}
#ChildThing: [Id=_]: {
  id: Id
  ...
}

things: [Id=_]: {
  id: Id
  children: #ChildThing
}

things: {
  thing_a: {
    children: {
      child_a: #ChildThingA & {
        attr_a: "foo"
      },
    },
  },
}

// The output we're looking for (limited to one "thing" and one "child" for succinctness):
// things: {
//   thing_a: {
//     children: {
//       child_a: {
//         id: "child_a",
//         attr_a: "foo",
//       },
//     },
//   },
// },

$ cue eval -c test.cue 
#ChildThingA.attr_a: conflicting values string and {id:Id,...} (mismatched types string and struct):
    ./test.cue:1:15
    ./test.cue:2:11
    ./test.cue:7:22
#ChildThingB.attr_b: conflicting values string and {id:Id,...} (mismatched types string and struct):
    ./test.cue:4:15
    ./test.cue:5:11
    ./test.cue:7:22
things.thing_a.children.child_a.attr_a: conflicting values string and {id:Id,...} (mismatched types string and struct):
    ./test.cue:1:15
    ./test.cue:2:11
    ./test.cue:7:22
    ./test.cue:20:16
things.thing_a.children.child_a.id: conflicting values "child_a" and {id:Id,...} (mismatched types string and struct):
    ./test.cue:1:15
    ./test.cue:7:22
    ./test.cue:8:7
    ./test.cue:20:16
```

**Comment by jlongtine** [comment, marked as answer]
@MattF-NSIDC I think templates are still throwing you off a bit here. 

I changed your code a bit and got the output you're looking for: 

```
#ChildThingA: #ChildThing & {
  attr_a: string
}
#ChildThingB: #ChildThing & {
  attr_b: string
}

#ChildThing: {
  id: string
  ...
}

things: [Id=_]: {
  id: Id
  children: [ChildId=_]: #ChildThing & {
    id: ChildId
  }
}

things: {
  thing_a: {
    children: {
      child_a: #ChildThingA & {
        attr_a: "foo"
      }
    }
  }
}
```

The main difference is that I changed where the template/bulk optional field is was defined. This changes `#ChildThing` from a struct with two levels, to a flat struct that simply has an `id` and is open. I hope this helps. Let me know if I can help you deepen your understanding of this!

**Reply by MattF-NSIDC** [reply]
This worked beautifully. I tried to click "Mark as answer", but nothing happened :confused: 

I see now how this override of the `#ChildThing` type works within the template, but I suppose I'm still struggling to understand the "lattice" concept. Is there any tooling that can generate a visualization of a lattice for a CUE program?

**Reply by jlongtine** [reply]
I'm not aware of any tooling to visualize the lattice. One thing to note is that the idea of the lattice is related to a specific value in CUE, rather than all of the values in a configuration. Granted, a "value" might be another struct. 

I think the best way I've come up to think about this is that `cue` provides tools to create a multi-level struct where each node in that struct can be constrained (by a type, like `string`, `int`, or another struct or list... or by a concrete value like `'foo'`, `1`, `[1,2,3]`... or by a disjunction of a combination of those things). 

The use of templates in your example is basically just saying: for any field that is defined here, it must take the shape `#ChildThing` (which has an `id` field). The use of `id: ChildId` forces the `id` to be the same string as the field name under `children`.

CUE is _all_ about constraints. Going from general to specific, at each level, all the way down to the individual values that get output.

**Reply by myitcv** [reply]
@MattF-NSIDC welcome to the CUE project!

> I tried to click "Mark as answer", but nothing happened 😕

Are you declaring all the `#ChildThingXYZ` definitions? Is this an exhaustive list?

Because if so the following might be a better approach:


```
#ChildThingA: {
  _#ChildThing
  attr_a: string
}

#ChildThingB: {
  _#ChildThing
  attr_b: string
}

_#ChildThing: {
  id: string
}

#ChildThing: #ChildThingA | #ChildThingB

things: [Id=_]: {
  id: Id
  children: [ChildId=_]: #ChildThing & {
    id: ChildId
  }
}

things: {
  thing_a: {
    children: {
      child_a: #ChildThingA & {
        attr_a: "foo"
      }
    }
  }
}
```

([playground](https://cuelang.org/play/?id=GvJBgtnTwz8#cue@export@cue))

Otherwise, you effectively lose type safety with respect to the children (I have two kids, and this doesn't sound like a good thing to lose!). 

Note the use of am embedding `_#ChildThing` which effectively represents the common parts of a child thing. Embedding of definitions is particularly useful, because the act of embedding disregards the closedness allowing you to define additional fields, then recursively closes the result (`#ChildThingA` or `#ChildThingB` in this case). See [the spec](https://tip.cuelang.org/docs/references/spec/#embedding) for more details.
